lib/pbench/server/api/resources/query_apis/metadata_index/__init__.py

```diff
@@ -0,0 +1,103 @@
+from http import HTTPStatus
+from logging import Logger
+from typing import AnyStr
+
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import JSON, Schema, SchemaError
+from pbench.server.api.resources.query_apis import CONTEXT, ElasticBase
+from pbench.server.database.models.datasets import Dataset, Metadata, MetadataError
+from pbench.server.database.models.users import User
+
+
+class MissingRunIdSchemaParameter(SchemaError):
+    """
+    The subclass schema is missing the required "run_id" parameter required
+    to locate a Dataset.
+    """
+
+    def __init__(self, subclass_name: str):
+        super().__init__()
+        self.subclass_name = subclass_name
+
+    def __str__(self) -> str:
+        return f"API {self.subclass_name} is missing schema parameter run_id"
+
+
+class RunIdBase(ElasticBase):
+    """
+    A base class for query apis that depends on Metadata for getting the
+    indices.
+
+    This class extends the ElasticBase class and implements a common
+    `preprocess` method based on client provided run id. The subsequent methods
+    such as 'assemble' and 'postprocess' need to be implemented by a respective
+    subclasses.
+
+    Note that run_id is a required schema parameter for all the
+    classes inheriting this class. Also "preprocess" provides json context
+    containing dataset and a run_id that's passed to the assemble and
+    postprocess methods.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger, schema: Schema):
+        if "run_id" not in schema:
+            raise MissingRunIdSchemaParameter(self.__class__.__name__)
+        super().__init__(config, logger, schema)
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        request is authorized for this dataset.
+
+        If the user has authorization to read the dataset, return the Dataset
+        object and run_id as JSON CONTEXT so that the postprocess operations
+        can use it to identify the index to be searched from document index
+        metadata.
+        """
+        run_id = client_json["run_id"]
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error(f"Dataset with Run ID {run_id!r} not found")
```

Comment:
Well this log message may not be helpful in the long run (like if we think this is not adding any value for debugging purpose and just taking up a log space then we can remove it later) but until then it might help us in debugging since currently our APIs are in active development mode. But I agree that this probably should't be an error message, info/debug would be better.

Reply:
This code depends on having a `Dataset` resource in order to identify the appropriate indices; and as our server code becomes increasingly database-driven, yes, I think it is an error. Now, if there are in fact Elasticsearch documents and a tarball with that `run_id` but it's somehow not in the `Dataset` table, that's arguably an internal error rather than a `NOT_FOUND`, but under normal circumstances it ought to be reasonable to assume that if there's no `Dataset` with the specified `run_id`, then the `run_id` was erroneous and the resource is `NOT_FOUND`.

Reply:
The erroneous `run_id` doesn't mean its an error on our side right so the log message can be debug right?

Reply:
"It's complicated" ... it's either a user-error because they made up a bad `run_id`, or it's an internal error because they had a *valid* `run_id` (probably from a previous Pbench query), but we can't find the `Dataset` resource. I suppose we really don't want to be returning an internal error on random bad input data, though; so "not found" on a very rare and unlikely internal error path might be the better of the two alternatives.

Reply:
I don't see how we can know the reason why a `run_id` is not found.  So logging this as an `error` is just going to create noise.  If the dashboard detects the inconsistency, then using a tool like Sentry, or LogRocket, the server can be notified about what is happening with the client.  But here there isn't enough information to know if a `run_id` should be found but isn't, or if the `run_id` never existed, or no longer exists.

Reply:
Updated to `debug`